021-server-side-rendering/021-server-side-rendering.md

```diff
@@ -0,0 +1,296 @@
+
+**Response**
+
+Session object.
+
+### Data Structure
+
+**Session Object Additions**
+
+| Name   | Type   | Description    |
+| ------ | ------ | -------------- |
+| secret | String | Session token. |
+
+### Documentation & Content
+
+#### What **docs** would support this feature?
+
+- SDK docs and examples for the new helper method.
+- SDK docs and examples for the new token exchange endpoint.
+
+#### What **tutorials** (text/video) might help developers understand this feature scope, capabilities, and possible use-cases?
+
+- Generic Tutorial for using Appwrite for SSR
+- Tutorial for using Appwrite and Next.js for SSR
+- Tutorial for using Appwrite and SvelteKit for SSR
+
+#### What **demo applications** can help us demonstrate this feature APIs and capabilities?
+
+- Update Almost SSR examples.
+
+### Prior art
+
+[prior-art]: #prior-art
+
+- https://supabase.com/docs/guides/auth/server-side-rendering
+- https://firebase.nuxtjs.org/tutorials/ssr.html
+- https://github.com/gladly-team/next-firebase-auth
+
+### Unresolved questions
+
+[unresolved-questions]: #unresolved-questions
+
+**Problem #1**
+
+1.1. Which solution should we implement? A or B?
+
+**Problem #2**
+
+2.1. Should we remove the undocumented workaround?
+
+2.2. Do we need to add a new endpoint to exchange the temporary token for a session token? Can we reuse the existing magic URL exchange endpoint?
+
+2.3. The Create OAuth2 Session endpoints are not included with Server Side SDKs. Should we include them? If so, server side apps need the Location header to forward to the user. How should we make this accessible from the SDK?
+
+**Problem #3**
+
+3.1. Should the helper method be available in Server Side SDKs? How do we seperate requests that are authenticated with a session token, and requests that are authenticated with an API key?
+
+**General**
+
+4.1. Incorporating some of these SSR changes reduces the distinction between the Server Side and Client Side SDKs. Which SDKs would be recommended for SSR applications?
+
+### Future possibilities
+
+[future-possibilities]: #future-possibilities
+
+**Cookie helper methods for SSR frameworks**
+
+On the web, SSR applications will need to set cookies under the SSR domain, but the cookie returned by Appwrite is set under the Appwrite domain.
+
+Each SSR framework has an interface for setting cookies in the response.
+For example, in Next.js, you can use the [setCookie](https://nextjs.org/docs/api-reference/next/cookies#setcookie) method.
+SvelteKit has a [cookie](https://kit.svelte.dev/docs#modules-cookie) module.
+
+For Next.js, the code would look like this:
+
+```js
+import { cookies } from "next/headers";
+
+const cookieList = cookies();
+cookieList.set({
+  name: "my_cookie_name",
+  value: session.token,
+  path: "/",
+  sameSite: "none",
+  secure: true,
+  httpOnly: true,
+  maxAge: session.expire,
+});
+```
+
+We can provide a helper method for popular SSR frameworks to set the cookie in the response.
```

Comment:
Love this! Would be great for adoption.

Reply:
Yeh, this is a must. Can we include it in the main SDK or do we have to create separated packages for it?

Reply:
Can be in the main SDK :)

Reply:
@loks0n What's the expected effect on package size, build time, etc.

If it's trivial we can add, but I wonder when we draw the line for framework specific baggage in the main SDK

Reply:
@gewenyu99 in general nowadays tree-shaking should deal with most issues regarding bundle size on your app's final output, build time too. The package size would be bigger on the `node-modules` folder though.

Alternatively, you can monorepo the SDK, and add adapters. E.g. `@appwrite.io/adapter-svelte`